src/urbit/lib/chess.hoon

```diff
@@ -472,28 +472,28 @@
     board
   ?-    file
        %i
-     $(brd t.brd, rank (chess-rank (sub rank 1)), file %a)
+     $(brd t.brd, rank ?:(=(rank %1) rank (chess-rank (sub rank 1))), file %a)
```

Comment:
Okay, I've made this parser a lot more stable by moving the casts to the very end. I've also fixed a bug that I made, which conflated the file and the position in the line that we're parsing. Replace lines `406` - `502` with this code:
```
::
::
+|  %parsing
::
::  turn fen into chess-position
++  fen-to-position
  |_  fen=chess-fen
  ++  $
    ^-  chess-position
    %+  rash
      fen
    ;~  (glue ace)
        fen-to-board         ::  board
        fen-to-player        ::  player-to-move
        fen-to-white-castle  ::  white-can-castle
        fen-to-black-castle  ::  black-can-castle
        fen-to-en-passant    ::  en-passant
        dem                  ::  ply-50-move-rule
        dem                  ::  move-number
    ==
  ::
  ::  XX: need to implement
  ::  turn fen into chess-position with checks for valid fen
  ++  safe
    ^-  chess-position
    ~&  %not-implemented  !!
  ++  fen-to-board
    %+  cook  board-helper
    %+  more  (just '/')
    ::  XX is this actually slower than alf and fail on invalid char?
    (stun [1 8] (mask "12345678bknpqrBKNPQR"))
  ++  fen-to-player
    ;~  pose
        (cold %white (just 'w'))
        (cold %black (just 'b'))
    ==
  ++  fen-to-white-castle
    ;~  pose
        (white-castle-helper hep " -" %none)
        (white-castle-helper (just 'k') " k" %none)
        (white-castle-helper (just 'q') " q" %none)
        (white-castle-helper (jest 'KQ') " " %both)
        (white-castle-helper (just 'K') " " %kingside)
        (white-castle-helper (just 'Q') " " %queenside)
    ==
  ++  fen-to-black-castle
    ;~  pose
        (cold %none hep)
        (cold %both (jest 'kq'))
        (cold %kingside (just 'k'))
        (cold %queenside (just 'q'))
    ==
  ++  fen-to-en-passant
    ;~  pose
        (cold ~ hep)
        %+  cook
            some
            ;~  plug
                (cook chess-file (cook term (shim 'a' 'h')))
                (cook chess-rank (cook term (mask "36")))
            ==
    ==
  ++  board-helper
    |=  brd=(list tape)
    ?>  =((lent brd) 8)
    =/  rank=@  %8
    =/  file=@  %a
    =/  pos=@  0
    =/  board  *chess-board
    |-
    ?~  brd
      board
    ?:  =(file %i)
      $(brd t.brd, rank (sub rank 1), file %a, pos 0)
::      $(brd t.brd, rank ?:(=(rank %1) rank (chess-rank (sub rank 1))), file %a)
    =*  next  ~&  i.brd  ~&  file  ~&  pos  (snag pos i.brd)
    ?:  (lte next 56)
      $(file (add file (sub next 48)), pos +(pos))
    ?:  (gte next 97)
      ?+  (term next)  !!
        %b  $(file +(file), pos +(pos), board (~(put by board) [[(chess-file file) (chess-rank rank)] [%black %bishop]]))
        %k  $(file +(file), pos +(pos), board (~(put by board) [[(chess-file file) (chess-rank rank)] [%black %king]]))
        %n  $(file +(file), pos +(pos), board (~(put by board) [[(chess-file file) (chess-rank rank)] [%black %knight]]))
        %p  $(file +(file), pos +(pos), board (~(put by board) [[(chess-file file) (chess-rank rank)] [%black %pawn]]))
        %q  $(file +(file), pos +(pos), board (~(put by board) [[(chess-file file) (chess-rank rank)] [%black %queen]]))
        %r  $(file +(file), pos +(pos), board (~(put by board) [[(chess-file file) (chess-rank rank)] [%black %rook]]))
      ==
    ?+  (term (add next 32))  !!
      %b  $(file +(file), pos +(pos), board (~(put by board) [[(chess-file file) (chess-rank rank)] [%white %bishop]]))
      %k  $(file +(file), pos +(pos), board (~(put by board) [[(chess-file file) (chess-rank rank)] [%white %king]]))
      %n  $(file +(file), pos +(pos), board (~(put by board) [[(chess-file file) (chess-rank rank)] [%white %knight]]))
      %p  $(file +(file), pos +(pos), board (~(put by board) [[(chess-file file) (chess-rank rank)] [%white %pawn]]))
      %q  $(file +(file), pos +(pos), board (~(put by board) [[(chess-file file) (chess-rank rank)] [%white %queen]]))
      %r  $(file +(file), pos +(pos), board (~(put by board) [[(chess-file file) (chess-rank rank)] [%white %rook]]))
  ==
  ++  white-castle-helper
    |*  [rul=rule pre=tape res=chess-castle]
    ;~(pfix rul (funk pre (easy res)))
--  --
```

The changes are actually quite small:
- Add a chapter for organization purposes
- Move `fen-to-position-safe` into `fen-to-position`, just like we did with `position-to-fen` and `simplified`
- Turn `fen-to-position` into a `door` with a default `arm`
- Add `pos` to `board-helper` to track what position in the FEN board line we're parsing
- Change the types of `rank` and `file` to `@`, and only cast them to `chess-rank` and `chess-file` (respectively) during insertion into the board object

Reply:
After rebasing and updating the parser I've narrowed down the issue to `fen-to-en-passant`. In the dojo I can get FENs with no possible en passant (`-`) to parse, but adding an en passant square causes a crash with the following stack trace. Going to come back to this later tonight.

```
/lib/chess/hoon:<[414 5].[425 7]>
/lib/chess/hoon:<[415 5].[425 7]>
/sur/chess/hoon:<[40 3].[40 29]>
/sur/chess/hoon:<[39 3].[40 29]>
/sur/chess/hoon:<[40 26].[40 28]>
dojo: hoon expression failed
```